MenuItem: Add right padding for unchecked radio and checkbox items

andrewserong opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
[Based on this suggestion](https://github.com/WordPress/gutenberg/pull/34375#issuecomment-908415069), this is an alternative to #34375 with the goal of ensuring consistent width of the dropdown menu items between checked and unchecked states.

This PR adds right padding to checkbox and radio dropdown menu items that are only children — that is, they do not have icons or shortcuts.

The goal is to ensure that the padding is accounted for in the menu item's unchecked state, so that when it is checked, the menu item takes up the same amount of room visually. It also ensures that unchecked menu items do not encroach upon the space used by icons to denote checked status.

The expected behaviour is most notable with the ToolsPanel when we add a longer menu item to it, so this PR also includes an update to the ToolsPanel story, for testing. See screenshots below:

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Manually using Storybook:

```
npm run storybook:dev
```

Go to http://localhost:50240/?path=/story/components-experimental-toolspanel--default and experiment with checking and unchecking the panel items.

## Screenshots <!-- if applicable -->

With other items checked (note there isn't enough whitespace next to the third item):

| Before | After |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/131295992-21a5269e-9304-4f22-88d4-041356348835.png) | ![image](https://user-images.githubusercontent.com/14988353/131296006-91003ac9-8829-4322-bcd7-53bab9a0ed82.png) |

With other items unchecked (note that in the before gif, the width of the popover is inconsistent between all items unchecked and one item checked)

| Before | After |
| --- | --- |
| ![tools-panel-menu-item-before-sml](https://user-images.githubusercontent.com/14988353/131296652-5ce48fc3-26d3-452e-aad1-8ebce247dc32.gif) | ![tools-panel-menu-item-after-sml](https://user-images.githubusercontent.com/14988353/131296688-7f55fcc0-4ec2-4c6b-a44b-9e0ef85c11d8.gif) |

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
(non-breaking change which fixes a styling issue)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

andrewserong committed:
MenuItem: Add right padding for unchecked radio and checkbox items

jasmussen commented:
Nice work, this is what I see:

![tools](https://user-images.githubusercontent.com/1204802/131472096-e6de25c8-f94a-44a9-a361-b52b165b1420.gif)

Those scrollbars, not sure why they are there. I suspect it's primarily something affecting the storybook rendering, though, so it may not need a fix. 

If it does need a fix, and doing a little inspecting, I noticed that something like the following would fix it up:

```
.components-popover__content > div {
	width: 100%;
}
```

ciampo commented:
> Those scrollbars, not sure why they are there. I suspect it's primarily something affecting the storybook rendering, though

Noting the same.

@andrewserong, is there an easy way to check if the scrollbars would disappear when the `ToolsPanel` is rendered inside the editor's sidebar?

On the other hand, I believe that @jasmussen 's suggested fix shouldn't hurt anyway.

jasmussen commented:
We'd probably want that width fix to be added to the popover component directly, though, in case that's what it takes.

jasmussen commented:
It appears the scrollbars is a regression in trunk as well 🙈

<img width="354" alt="Screenshot 2021-08-31 at 14 55 29" src="https://user-images.githubusercontent.com/1204802/131506241-41107077-e865-4fb9-8363-efb42cb97397.png">

So it seems we need a fix. Whether slipstreamed in this one, or separately.

andrewserong commented:
Thanks for reviewing @jasmussen and @ciampo, I was originally working on another branch, so didn't notice the scrollbars when I moved the change to a fresh branch from trunk.

It looks like the regression was introduced in https://github.com/WordPress/gutenberg/pull/34378 — by reducing the padding on the div child of `.components-popover__content`, several other components now use negative margins that are too large. In this case, it's the `.components-menu-group` that sets each of the margins to negative `-$grid-unit-15` instead of the now expected `-$grid-unit-10`. I've updated that component in this PR now, too.

@jasmussen was the goal with #34378 to update all instances of the popover content to the smaller padding? If so, we can work through the list of instances of `-$grid-unit-15` and update to `-$grid-unit-10` across a separate PR(s)?

andrewserong commented:
> is there an easy way to check if the scrollbars would disappear when the ToolsPanel is rendered inside the editor's sidebar?

Just to add, for testing scrollbars within the editor, we should be able to do that with the Dimensions control on the Group block. The Reset all button receives the padding rule from this PR:

![image](https://user-images.githubusercontent.com/14988353/131605180-1f8d8735-9171-4682-a3d5-326728e02910.png)

jasmussen commented:
> @jasmussen was the goal with #34378 to update all instances of the popover content to the smaller padding? If so, we can work through the list of instances of -$grid-unit-15 and update to -$grid-unit-10 across a separate PR(s)?

Yes, my mistake, I'll spin up a PR for that now, sorry about that!

andrewserong commented:
No worries, thanks for coming up with a quick fix! 🙇

andrewserong commented:
I've just given this a rebase, and it appears to be working well now with no scrollbars. Let me know if you'd like any changes!

jasmussen commented on packages/components/src/menu-item/style.scss:
```diff
@@ -2,6 +2,13 @@
+	&[role="menuitemradio"],
+	&[role="menuitemcheckbox"] {
+		.components-menu-item__item:only-child {
+			padding-right: $grid-unit-60;
```
Should we add a comment to denote what this particular value does?

andrewserong replied:
Added 👍

jasmussen reviewed (approved):
Looks good, nice work.

One thing I noticed is that now we've removed the min-width of popover menus, the tools panel is rather small. As we should for all popovers on a case by case basis, we can add back min-widths where it makes sense. I feel like to makes sense here, and that we should consider adding a min-width: 200px; to this particular dropdown. Feel free to do that here if you like, or we can follow up separately.

aaronrobertshaw reviewed (approved):
LGTM 👍 

✅ ToolsPanel menu items display neatly regardless of checked status.

<img width="387" alt="Screen Shot 2021-09-03 at 2 31 15 pm" src="https://user-images.githubusercontent.com/60436221/131951383-eabc50ce-62a2-4784-974c-002ece160fea.png">

andrewserong committed:
Add comment explaining the reason for the padding value

andrewserong commented:
Thanks folks, I've added in a comment explaining the reason for the padding. For the sake of neatness in the changelog, I'll add the minimum width for the ToolsPanel dropdown in a separate PR, and merge this one in shortly.

andrewserong commented:
The e2es were failing, preventing a merge, so I've just given this a rebase.

andrewserong commented:
>  I'll add the minimum width for the ToolsPanel dropdown in a separate PR

I ran out of time to spin up a separate PR for that today, but will take a look next week if no-one beats me to it 🙂

aaronrobertshaw commented:
> I ran out of time to spin up a separate PR for that today, but will take a look next week if no-one beats me to it 🙂

We can add the min-width via the `ToolsPanelHeader` dropdown menu's `menuProps`. I've trialed this approach within a [draft PR](https://github.com/WordPress/gutenberg/pull/34530) refining the `ToolsPanel` UX. https://github.com/WordPress/gutenberg/commit/3462e40a36a83c727d869290659317aa86ac81b8

ciampo commented:
> As we should for all popovers on a case by case basis, we can add back min-widths where it makes sense.

Should we check and potentially add a `min-width` for the rest of the popovers too?

jasmussen commented:
> Should we check and potentially add a min-width for the rest of the popovers too?

I created #34532 which might provide a second layer fallback, which provides back a min-width if you use menu items inside.

andrewserong commented:
Thanks Aaron and Joen, between those two approaches, it looks like we're covered for the min-width 👍

aaronrobertshaw commented:
> between those two approaches, it looks like we're covered for the min-width 👍

The application of a `200px` min-width on the ToolsPanel menu is superfluous now with the inner menu item getting a `160px` min-width. The menu item icon plus its margin (or this PR's padding) push it out past `200px` so for now I think we can just rely on #34532

andrewserong commented:
Ah, of course, that makes sense, sounds good.

aaronrobertshaw commented:
Actually, don't mind me. If the ToolsPanel menu only had optional controls and none are selected then we could get inconsistent menu widths. So a min-width on the ToolsPanel menu probably still makes sense.

| Before Selection | After Selection |
|---|---|
| <img width="282" alt="Screen Shot 2021-09-06 at 2 58 30 pm" src="https://user-images.githubusercontent.com/60436221/132162976-9bd1b34d-5676-4950-bc46-26637848c081.png"> | <img width="283" alt="Screen Shot 2021-09-06 at 2 58 35 pm" src="https://user-images.githubusercontent.com/60436221/132162990-a640015f-c5df-4926-aefd-b662fbc99776.png"> |
